File: 2. SVD Decomposition.cpp
// Topic: SVD Decomposition using Eigen library (note: requires Eigen and linking)
// This file demonstrates usage; Eigen must be available to compile.
#include <bits/stdc++.h>
#if __has_include(<Eigen/Dense>)
#include <Eigen/Dense>
using namespace Eigen;
int main()
{
    MatrixXd A(3, 2);
    A << 1, 2, 3, 4, 5, 6;  // 3x2 matrix
    JacobiSVD<MatrixXd> svd(A, ComputeThinU | ComputeThinV);
    cout << "Singular values:\n" << svd.singularValues() << '\n';
    // reconstruct using top k
    int k = 1;
    MatrixXd U = svd.matrixU().leftCols(k);
    MatrixXd V = svd.matrixV().leftCols(k);
    VectorXd S = svd.singularValues().head(k);
    MatrixXd Arec = U * S.asDiagonal() * V.transpose();
    cout << "Reconstructed:\n" << Arec << '\n';
    return 0;
}
#else
int main()
{
    std::cout << "Eigen not installed. To use this file, install Eigen and compile accordingly."
              << std::endl;
    return 0;
}
#endif

// Time Complexity: O(min(m,n)^2 * max(m,n)) depending on implementation
// Space Complexity: O(m*n)

File: 2. SVD Decomposition.py
# Topic: SVD Decomposition using numpy.linalg.svd and reconstruct using top-k

import numpy as np

def svd_reconstruct(A, k):
    U, S, Vt = np.linalg.svd(A, full_matrices=False)
    S_k = np.diag(S[:k])
    Arec = U[:,:k] @ S_k @ Vt[:k,:]
    return Arec, S

if __name__ == '__main__':
    A = np.array([[1,2],[3,4],[5,6]], dtype=float)
    Arec, S = svd_reconstruct(A, k=1)
    print('Singular values:', S)
    print('Reconstructed with k=1:\n', Arec)

# Time Complexity: O(min(m,n)^2 * max(m,n))
# Space Complexity: O(m*n)
